url39209c0557_au_research_ANDS_RIFCS_API_tgz-pJ8-org_ands_rifcs_example_SampleRIFCSBuildCollectionV1Dot6Number1J8/src/org/ands/rifcs/base/RIFCSWrapper.java

```diff
@@ -0,0 +1,371 @@
+            StringWriter writer = new StringWriter();
+            Result result = new StreamResult(writer);
+            Source source = new DOMSource(doc);
+            transformer.transform(source, result);
+            writer.close();
+            return writer.toString();
+        } catch (TransformerException te) {
+            return null;
+        } catch (IOException ioe) {
+            return null;
+        }
+        //DOMImplementation impl = doc.getImplementation();
+        //DOMImplementationLS implLS =
+        //  (DOMImplementationLS) impl.getFeature("LS","3.0");
+        //LSSerializer writer = implLS.createLSSerializer();
+        // This is to suppress the xml header, with version
+        // and the encoding being automatically generated
+        //writer.getDomConfig().setParameter("xml-declaration", false);
+        //return writer.writeToString(doc);
+    }
+
+
+    /**
+     * Validate against the most recent rif-cs schema. The schema is
+     * accessed remotely from the production site. If wanting to use a
+     * local cached schema use the other validate method.
+     *
+     * @throws SAXException
+     *      if document is invalid
+     * @throws MalformedURLException
+     *      if schema URL is invalid
+     * @throws IOException
+     *      if URL stream cannot be accessed
+     * @throws ParserConfigurationException
+     *      A ParserConfigurationException
+     */
+    public final void validate() throws SAXException,
+    MalformedURLException, IOException, ParserConfigurationException {
+        // create a SchemaFactory capable of understanding WXS schemas
+        SchemaFactory factory = SchemaFactory.newInstance(
+                XMLConstants.W3C_XML_SCHEMA_NS_URI);
+
+        Schema schema = factory.newSchema(doXercesWorkaround());
+
+        // create a Validator instance, which can be used
+        // to validate an instance document
+        Validator validator = schema.newValidator();
+        validator.validate(new DOMSource(doc));
+    }
+
+    /** Xerces cannot handle multiple schema files with the same namespace,
+     *  so we need to work around this.
+     *
+     * @return
+     *      The RIF-CS schema in a form usable by Xerces.
+     *
+     * @throws SAXException
+     *      if document is invalid
+     * @throws MalformedURLException
+     *      if schema URL is invalid
+     * @throws IOException
+     *      if URL stream cannot be accessed
+     * @throws ParserConfigurationException
+     *      A ParserConfigurationException
+    */
+    private Source doXercesWorkaround() throws SAXException,
+    MalformedURLException, IOException, ParserConfigurationException {
+        DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
+        factory.setNamespaceAware(true);
+        DocumentBuilder builder = factory.newDocumentBuilder();
+
+        Document docRO = builder.parse(new URL(
+                Constants.SCHEMA_REGISTRY_OBJECTS).openStream());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource returned by the following method call: *openStream*. Currently, there are execution paths that do not contain closure statements, for example, when DocumentBuilder.parse throws an exception. Ensure the object returned by *openStream()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).